File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/encryption/secrets/UserAuthenticationPasswordSpec.kt
package stasis.test.client_android.lib.encryption.secrets

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.shouldBe
import okio.ByteString.Companion.encodeUtf8
import stasis.client_android.lib.encryption.secrets.UserAuthenticationPassword
import java.util.UUID

class UserAuthenticationPasswordSpec : WordSpec({
    "A Hashed UserAuthenticationPassword" should {
        "allow extracting the hashed password" {
            val originalPassword = "test-password"
            val expectedPassword = "dGVzdC1wYXNzd29yZA"

            val actualPassword = UserAuthenticationPassword.Hashed(
                user = UUID.randomUUID(),
                hashedPassword = originalPassword.encodeUtf8()
            )

            actualPassword.extract() shouldBe (expectedPassword)
        }

        "fail if the password is extracted more than once" {
            val originalPassword = "test-password"
            val expectedPassword = "dGVzdC1wYXNzd29yZA"

            val actualPassword = UserAuthenticationPassword.Hashed(
                user = UUID.randomUUID(),
                hashedPassword = originalPassword.encodeUtf8()
            )

            actualPassword.extract() shouldBe (expectedPassword)

            shouldThrow<IllegalStateException> {
                actualPassword.extract()
            }
        }

        "support providing a digested version of the hashed password" {
            val originalPassword = "test-password"
            val digestedPassword =
                "c2KtMdtdEMnxjhRjqxjdvpxUgZzO410G0LK09qrg-zqMDL9qSaIgoLRoU2X7ueHOa4_XWAcABv4lXD6B0ddUwA"

            val actualPassword = UserAuthenticationPassword.Hashed(
                user = UUID.randomUUID(),
                hashedPassword = originalPassword.encodeUtf8()
            )

            actualPassword.digested() shouldBe (digestedPassword)
        }
    }

    "A Unhashed UserAuthenticationPassword" should {
        "allow extracting the raw password" {
            val originalPassword = "test-password"

            val actualPassword = UserAuthenticationPassword.Unhashed(
                user = UUID.randomUUID(),
                rawPassword = originalPassword.encodeUtf8()
            )

            actualPassword.extract() shouldBe (originalPassword)
        }

        "fail if the password is extracted more than once" {
            val originalPassword = "test-password"

            val actualPassword = UserAuthenticationPassword.Unhashed(
                user = UUID.randomUUID(),
                rawPassword = originalPassword.encodeUtf8()
            )

            actualPassword.extract() shouldBe (originalPassword)

            shouldThrow<IllegalStateException> {
                actualPassword.extract()
            }
        }

        "support providing a digested version of the raw password" {
            val originalPassword = "test-password"
            val digestedPassword =
                "c2KtMdtdEMnxjhRjqxjdvpxUgZzO410G0LK09qrg-zqMDL9qSaIgoLRoU2X7ueHOa4_XWAcABv4lXD6B0ddUwA"

            val actualPassword = UserAuthenticationPassword.Unhashed(
                user = UUID.randomUUID(),
                rawPassword = originalPassword.encodeUtf8()
            )

            actualPassword.digested() shouldBe (digestedPassword)
        }
    }
})

File: client/src/test/scala/stasis/test/specs/unit/client/encryption/secrets/UserAuthenticationPasswordSpec.scala
package stasis.test.specs.unit.client.encryption.secrets

import org.apache.pekko.util.ByteString

import stasis.client.encryption.secrets.UserAuthenticationPassword
import stasis.shared.model.users.User
import stasis.test.specs.unit.UnitSpec

class UserAuthenticationPasswordSpec extends UnitSpec {
  "A Hashed UserAuthenticationPassword" should "allow extracting the hashed password" in {
    val originalPassword = "test-password"
    val expectedPassword = "dGVzdC1wYXNzd29yZA"
    val actualPassword = UserAuthenticationPassword.Hashed(
      user = User.generateId(),
      hashedPassword = ByteString(originalPassword)
    )

    actualPassword.extract() should be(expectedPassword)
  }

  it should "fail if the password is extracted more than once" in {
    val originalPassword = "test-password"
    val expectedPassword = "dGVzdC1wYXNzd29yZA"
    val actualPassword = UserAuthenticationPassword.Hashed(
      user = User.generateId(),
      hashedPassword = ByteString(originalPassword)
    )

    actualPassword.extract() should be(expectedPassword)

    an[IllegalStateException] should be thrownBy actualPassword.extract()
  }

  it should "support providing a digested version of the hashed password" in {
    val originalPassword = "test-password"
    val digestedPassword = "YrlMMruc2Em28w6fWV8YR9CYQYO4twcBhvWdczLS-9MjJchod8u4iUTbxVTpVOojGn3iCDXJpLHyNRXfR15kuQ"
    val actualPassword = UserAuthenticationPassword.Hashed(
      user = User.generateId(),
      hashedPassword = ByteString(originalPassword)
    )

    actualPassword.digested() should be(digestedPassword)
  }

  "An Unhashed UserAuthenticationPassword" should "allow extracting the raw password" in {
    val originalPassword = "test-password"
    val actualPassword = UserAuthenticationPassword.Unhashed(
      user = User.generateId(),
      rawPassword = ByteString(originalPassword)
    )

    actualPassword.extract() should be(originalPassword)
  }

  it should "fail if the password is extracted more than once" in {
    val originalPassword = "test-password"
    val actualPassword = UserAuthenticationPassword.Unhashed(
      user = User.generateId(),
      rawPassword = ByteString(originalPassword)
    )

    actualPassword.extract() should be(originalPassword)

    an[IllegalStateException] should be thrownBy actualPassword.extract()
  }

  it should "support providing a digested version of the raw password" in {
    val originalPassword = "test-password"
    val digestedPassword = "YrlMMruc2Em28w6fWV8YR9CYQYO4twcBhvWdczLS-9MjJchod8u4iUTbxVTpVOojGn3iCDXJpLHyNRXfR15kuQ"
    val actualPassword = UserAuthenticationPassword.Unhashed(
      user = User.generateId(),
      rawPassword = ByteString(originalPassword)
    )

    actualPassword.digested() should be(digestedPassword)
  }
}
